h2o-core/src/main/java/water/rapids/ast/prims/reducers/AstTopN.java

```diff
@@ -0,0 +1,214 @@
+            "values.  If getBottomN is 1, we will return the bottom N percent.  If getBottomN is 0, we will return" +
+            "the top N percent of rows";
+  }
+
+  @Override
+  public ValFrame apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+    Frame frOriginal = stk.track(asts[1].exec(env)).getFrame(); // get the 2nd argument and convert it to a Frame
+    int colIndex = (int) stk.track(asts[2].exec(env)).getNum();     // column index of interest
+    double nPercent = stk.track(asts[3].exec(env)).getNum();        //  top or bottom percentage of row to return
+    int getBottomN = (int) stk.track(asts[4].exec(env)).getNum();   // 0, return top, 1 return bottom percentage
+    int totColumns = frOriginal.numCols();
+    long numRows = Math.round(nPercent*0.01*frOriginal.numRows()); // number of rows to return
+
+    // check for valid input parameter values
+    assert ((nPercent>0) && (nPercent<=100)); // make sure percent is between 0 and 100.
+    assert (numRows >= 0);                    // make sure we have at least one row to extract
+    assert ((colIndex >=0) && (colIndex < totColumns)); // valid column index specification
+    assert ((getBottomN ==0) || (getBottomN==1));       // must be 0 or 1
+    assert frOriginal.vec(colIndex).isNumeric();        // make sure we are dealing with numerical column only
+
+    String[] finalColumnNames = {"Original_Row_Indices", frOriginal.name(colIndex)}; // set output frame names
+    GrabTopN grabTask = new GrabTopN(finalColumnNames, numRows, (getBottomN==0));
+    grabTask.doAll(frOriginal.vec(colIndex));
+    return new ValFrame(grabTask._sortedOut);
+  }
+
+  /*
+   Here is the plan:
+   1. For each chunk (inside each map function of our MRTask), put all elements into a sorted heap with key as the
+    column value to grab and original row indices (in an array to deal with duplicated values) as the value.  At
+    the end of map, reduce the sorted heap to size N percent rows.
+   2. Inside reduce, make sure you combine the heaps and again reduce the sorted heap back to size N percent rows.
+   3. Inside the postGlobal, copy the heap (key, values).
+   */
+// E depends on column type: long or other numerics
+  public  class GrabTopN<E extends Comparable<E>> extends MRTask<GrabTopN<E>>  {
+    final String[] _columnName;   // name of column that we are grabbing top N for
+    TreeMap _sortHeap;
+    Frame _sortedOut;   // store the final result of sorting
+    final long _rowSize;   // number of top or bottom rows to keep
+    final boolean _increasing;  // sort with Top values first if true.
+    boolean _csLong=false;      // chunk of interest is long
+
+    private GrabTopN(String[] columnName, long rowSize, boolean increasing) {
+      _columnName = columnName;
+      _rowSize = rowSize;
+      _increasing = increasing;
+    }
+
+    @Override public void map(Chunk cs) {
+      _sortHeap = new TreeMap();
```

Comment:
I would use PriorityQueue instead. Priority queue uses heap instead of balanced tree